angr/sim_type.py

```diff
@@ -1187,23 +1189,15 @@ def _decl_to_type(decl, extra_types=None):
             elif key in ALL_TYPES:
                 struct = ALL_TYPES[key]
             else:
-                struct = None
+                extra_types[key] = struct
 
-            if struct is None or not struct.fields:
-                struct = SimStruct(fields, decl.name)
-            elif fields and struct.fields != fields:
-                raise ValueError("Redefining body of " + key)
+        if decl.decls is not None:
+            struct.fields = OrderedDict((field.name, _decl_to_type(field.type, extra_types)) for field in decl.decls)
```

Comment:
This allows mutation of existing structs when they are defined multiple times. I don't think that's valid. I'm pretty confident in the way this code was written originally. Do tests pass without the edits to `_decl_to_type`?

Reply:
Ah, I see. I think there's a better way to do this, though. Give me a second...

Reply:
Will this handle correctly self-referential unions and structs?

If the idea is avoid modify stored types, I can turn `if key in ALL_TYPES:` into `if key in ALL_TYPES and decl.decls is None:`. Maybe this can avoid the modification.

Reply:
Yes, it correctly handles self-referential unions and structs. Take a look at my new commit.

Your code modifies stored types too! If I stick a breakpoint after the struct parsing in `test_union_struct_referencing_each_other`, I can see that the structure in ALL_TYPES now has fields. This happens with and without my commit.

Reply:
fwiw I think modifying stored types is a good thing if they don't have any fields beforehand and this gives them fields

Reply:
I agree with you!

Reply:
Thanks for the help!